File: ackermann.py
import sys
import time

def ack(m, n):
    if m == 0:
        ans = n+1
    elif n == 0:
        ans = ack(m-1, 1)
    else:
        ans = ack(m-1, ack(m, n-1))
    return ans

# Python has a stack limit. By default its quite concervative. In order to do 4,1 in Ackermann need to increase the limit.
sys.setrecursionlimit(50000)

start = time.time()
answer = ack(4,1)
end = time.time()
print('Answer: %d Time: %d' % (answer, end-start))

File: ackermann.jl
import Pkg; 
Pkg.add("Memoize")

function ack(m, n)
    if m == 0
        ans = n + 1
    elseif n == 0
        ans = ack(m-1, 1)
    else
        ans = ack(m-1, ack(m, n-1))
    end
    return ans
end

using Memoize
@memoize ack3(m::Integer, n::Integer) = m == 0 ? n + 1 : n == 0 ? ack3(m - 1, 1) : ack3(m - 1, ack3(m, n - 1))

# For benchmarking.
@time begin
    println("ack results: ")
    ack(4, 1)
end

@time begin
    println("ack3 results: ")
    ack3(4, 2)
end
